File: backend/app/schemas/queue.py
"""Queue-related Pydantic schemas."""
import uuid
from datetime import datetime
from decimal import Decimal
from typing import Optional
from pydantic import BaseModel

from app.models.queue_entry import QueueStatus


class CustomerBrief(BaseModel):
    id: uuid.UUID
    phone: str
    name: Optional[str] = None
    is_vip: bool


class ServiceBrief(BaseModel):
    id: uuid.UUID
    name_ar: str
    name_en: str


class QueueEntryResponse(BaseModel):
    id: uuid.UUID
    position: int
    status: QueueStatus
    eta_minutes: float
    customer: CustomerBrief
    service: Optional[ServiceBrief] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class QueueEntryDetail(BaseModel):
    id: uuid.UUID
    tenant_id: uuid.UUID
    customer_id: uuid.UUID
    service_id: Optional[uuid.UUID] = None
    status: QueueStatus
    position: Optional[int] = None
    called_at: Optional[datetime] = None
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    amount: Optional[Decimal] = None
    notes: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class AddToQueueRequest(BaseModel):
    customer_id: uuid.UUID
    service_id: Optional[uuid.UUID] = None


class FinishServiceRequest(BaseModel):
    amount: Optional[float] = None
    notes: Optional[str] = None


class QueueStatsResponse(BaseModel):
    total_waiting: int
    total_called: int
    total_in_service: int
    avg_wait_time_minutes: float
    is_accepting_queue: bool

File: frontend/src/api/queue.ts
import api from './axios'

export type QueueStatus = 'waiting' | 'called' | 'in_service' | 'done' | 'cancelled' | 'no_show'

export interface QueueEntry {
  id: string
  position: number
  status: QueueStatus
  eta_minutes: number
  customer: { id: string; phone: string; name: string | null; is_vip: boolean }
  service: { id: string; name_ar: string; name_en: string; name_ru: string | null } | null
  created_at: string
}

export interface QueueEntryDetail {
  id: string
  tenant_id: string
  customer_id: string
  service_id: string | null
  status: QueueStatus
  position: number | null
  called_at: string | null
  started_at: string | null
  finished_at: string | null
  amount: number | null
  notes: string | null
  created_at: string
}

export interface QueueStats {
  total_waiting: number
  total_called: number
  total_in_service: number
  avg_wait_time_minutes: number
  is_accepting_queue: boolean
}

export const queueApi = {
  getQueue: () => api.get<QueueEntry[]>('/queue/').then((r) => r.data),
  getStats: () => api.get<QueueStats>('/queue/stats').then((r) => r.data),

  addToQueue: (data: { customer_id: string; service_id?: string }) =>
    api.post<QueueEntryDetail>('/queue/add', data).then((r) => r.data),

  callNext: () => api.post<QueueEntryDetail>('/queue/call-next').then((r) => r.data),

  startService: (entryId: string) =>
    api.post<QueueEntryDetail>(`/queue/${entryId}/start`).then((r) => r.data),

  finishService: (entryId: string, data: { amount?: number; notes?: string }) =>
    api.post<QueueEntryDetail>(`/queue/${entryId}/finish`, data).then((r) => r.data),

  cancel: (entryId: string) =>
    api.post<QueueEntryDetail>(`/queue/${entryId}/cancel`).then((r) => r.data),

  noShow: (entryId: string) =>
    api.post<QueueEntryDetail>(`/queue/${entryId}/no-show`).then((r) => r.data),

  toggleAccepting: () =>
    api.patch<{ is_accepting_queue: boolean }>('/queue/toggle-accepting').then((r) => r.data),
}
